### DISCUSSION TITLE
API routes and getServerSideProps timeout on production (Vercel)

### DISCUSSION THREAD

**Discussion by Pardaramskha** [vercel/next.js#47237] [Help] [answered]
### Summary

Recently, without any changes nor new deployments besides upgrade to NextJS 13, **API routes and SSG pages (with getServerSideProps) went dark**. It throws a **504 timeout error**, for both getServerSideProps functions and simple API routes from my functional components. 

Whether calls go through the API or not, it seems that any dynamic call that goes outsite of the app just throws a 504.

This only occurs in production (both main and sandbox) - does not occur in `dev` or `local build` environment. 

Vercel throws this:
```
504: GATEWAY_TIMEOUT
Code: FUNCTION_INVOCATION_TIMEOUT
```

Investigation on similar issues showed that it may be linked to different causes:
* [MUI Icons increasing bundle size](https://github.com/mui/material-ui/issues/35450) (but even when removing all @mui/icons-material from said pages, they still throws errors)
* Root directory position issue (but it seems it only occurs when deployed elsewhere than Vercel) 

The console won't show me any details about the timeout.

I'm **not** using the new NextJS architecture, and still rely on `/pages/api` for API routing.
Most of my functions use `/api/foo` for calling API logic.

Project config:
```
NextJS: 13.2.1
React: 18.2.0
MUI/MUI icons: 5.8.0
```

```
Deployed with: Vercel (hobby) - CI/CD from github
Node version: 16.X
Build commands: `yarn build && yarn postbuild` (postbuild used for next-sitemap)
Database: MongoDB atlas
```

Help is much appreciated !

**Comment by icyJoseph** [comment]
Hi,

So you call your API routes from within `GSSP`? 

- https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes

> It can be tempting to reach for an [API Route](https://nextjs.org/docs/api-routes/introduction) when you want to fetch data from the server, then call that API route from getServerSideProps. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both getServerSideProps and API Routes running on the server.

That's already a red flag for me. From within GSSP, you are already in an, specialized, API route (sort-of~~), so you should just get the data you need right here. 

Now, the FUNCTION_INVOCATION_TIMEOUT kicks in at 10 seconds without response.

- https://vercel.com/docs/concepts/limits/overview#general-limits

If it is the case that you are reaching to API routes from GSSP, then we are looking at debugging 2 round trips, instead of one.

I'd say, use `@next/bundle-analyzer`, and see the sizes and contents of the server bundle. 

If your bundles are large, and slow to parse, for example a lot of JSON config files, etc, then your time budget would be:

- parse files, including JSON assets
- evaluate scripts
- execute the code

Basically, downloading/parsing/compiling, but server side. Remember that these are all serverless lambdas, which have cold-starts.

If you are reaching to API routes from GSSP, you are, at worst, paying this, times two, but not necessarily.

Happy to bounce back ideas to get to the bottom of this.

**Reply by Pardaramskha** [reply]
Thank you for your answer!

To react on that:
* I don't use API calls within GSSP, I just mentionned that the bug occurs on both GSSP and API calls (the least being within functional components or page components) - I'll edit the initial post to make it clearer. 
* I've already used `@next/bundle-analyzer`, and the biggest of them all was `@mui/icons-material` - but even without calling it on page, the bug still occurs... I should try to completely remove the dependency or, at least, avoid relying on tree-shaking to import components by groups (i.e. `import { Menu, Facebook, Instagram } from "@mui/icons-material"`), but then I won't understand why it would work locally and not in production 😐

**Reply by icyJoseph** [reply]
Alright, but then the issue has to be something specific to the serverless deployment.

More precisely, either it is falling into an infinite loop and never resolves, or there's an inherently slow request happening.

What do these do? GSSP/API routes? Generally speaking, no need to be specific.

--- 

Are you sure you've set all env vars properly on the Vercel dashboard?

**Reply by Pardaramskha** [reply]
Generally, **GSSP fetches data from MongoDB** for posts and articles in order to hydrate `next-seo` tags and build dynamic SEO.
On the other hand, **API calls just do a simple CRUD routine**, fetching posts in order to display them, sending JSON to save a new post, etc...

To show some basic examples of how it's written:
```javascript
export async function getServerSideProps({ params }) {

    const clientPromise = await getClient()

    const post = await clientPromise
        .db(process.env.NEXT_PUBLIC_DATABASE_LABEL)
        .collection("articles")
        .findOne({ 'titleSlug': params.slug }, {projection: { "_id": 0 }})

    if (post) {
        const author = await clientPromise
            .db(process.env.NEXT_PUBLIC_DATABASE_LABEL)
            .collection("auteurs")
            .findOne({ 'slug': post.authorSlug }, {projection: { "_id": 0 }})

        await clientPromise.close()
        return { props: { post: post, author: author}}
    }

    await clientPromise.close()
    return { props: { error: true }}

}
```

About the env variables, they didn't move an inch - still the same as my .env.local file, proving that it's not some weird MongoDB "I forgot your access key" thing. 

Need to mention that it occurs, throwing me a 504, even with some dumb API call like so:

```javascript
// /pages/api/dumb.js

export default async function handler(req, res) {

    if(req.method !== 'GET') {
        return
    }

    console.log("john doe")

    res.status(200).json({})

}
```

**Reply by icyJoseph** [reply]
And no redirect/rewrite logic is colliding? Have you got a top level API folder? 

I've seen something like this in this forum before, on my phone now, but maybe you can search around here.

**Reply by Pardaramskha** [reply]
No, I don't have any rewrite logic anywhere...

My API folder is located in `[project]/pages` (this is why I can call it directly by fetching at `/api/[endpoint]`

I'm going to investigate it further, but from what I saw, similar problems were restricted to either GSSP or API calls, not both 🤔 

I'm wondering if this could be related to upgrading to Next 13 - maybe I've missed something?

**Reply by Pardaramskha** [reply]
Could it be related to my `jsconfig.json` ? I've recently added aliases for imports (but it worked fine for some days before going nuts)

```json
{
  "compilerOptions": {
    "baseUrl": "./src",
    "paths": {
      "@components/*": ["./components/*"],
      "@context": ["./context"],
      "@hooks": ["./hooks"],
      "@layouts": ["./layouts"],
      "@themes": ["./themes"]
    }
  }
}
```

**Comment by Pardaramskha** [comment]
I tried searching (again) for the solution within the discussions, but nothing has been answered yet:
* [Logging everything does not work](https://github.com/vercel/next.js/discussions/45734), for the API logic isn't even reached.
* [Re-building as in this post](https://github.com/vercel/next.js/discussions/44072) did not do the trick for me...
* [Seems to have no way to debug endpoint themselves](https://github.com/vercel/next.js/discussions/17119)

This [has recently been labelled as an issue](https://github.com/vercel/next.js/issues/47101) (waiting for similarities confirmation)

**Comment by Fasa2509** [comment]
I'm having the same problem. My app was working as expected three weeks ago, but then suddenly it just broke. Any api call to my own next endpoints takes +10s and returns 504 (and sometimes 502) errors, these calls also use a lot of memory, this includes too getServerSideProps.
I have tried to deploy in Netlify, use another Next version, fetching data with other frameworks, but nothing works.

I got this error message in my production logs

info  - Loaded env from /var/task/.env

**Comment by Pardaramskha** [comment, marked as answer]
I GOT IT, I finally understood what was wrong, and it's vicious (and have nothing to do with Next itself, in my case).

The issue came from Material UI, and specifically mui/icons-material.

I mostly used tree-shaking to import icons in my components (i.e. `import {AddRounded, Delete} from "@mui/icons-material"`) but the thing is doing so multiplies bundle size and build time by 5 to 6.

The vicious thing is that, in order to "remove" this abomination, we have to remove every single tree-shaking import from @mui/icons-material from the project, because the bug happens at build time. You can still have icons, but have to import them directly like so:
`import AddRounded from "@mui/icons-material/AddRoundedIcon"`

In my case, doing so made my builds go from 3-4mins. to merely 45seconds, and evaporated the bug.

You won't be able to fix this if you don't remove every single tree-shaking from MUI icons.

@Fasa2509 if you're not using MUI, maybe you should look into a bundle-analyzer like next-bundle-analyzer, or use a VSCode extension like "import size" (I don't remember the name) to watch for imports sizes in your files.

I'll close the discussion with that, but feel free to reopen it if your situation looks similar enough for this to have an impact on debugging it!

**Reply by theleewise** [reply]
THANK YOU!!!
This is exactly what was going on on 2 of my sites. What's dumb is I would have known this had I actually read the docs and not just grabbed the code and thought "yeah yeah, i know what I'm doing"

**Reply by pmuens** [reply]
I can confirm that using direct imports for MUI icons solved my problems too.

**Reply by sjsaul1101** [reply]
Thank you, solved my issue as well

**Reply by mihaben** [reply]
Ohh thank you! It solved my problem, I'd been 3 hours trying to solve this ❤️

**Reply by haleybe** [reply]
Solved the problem. In the entire project i had only 1 icon imported this way, and it was not even used. Spent several days so many hours trying to figure this out. Changed from server to static back and forth countless times. This is not just vicious, this comes from a deepest whole in hell, malevolent, ferocious, barbaric.  Thanks! ❤️

**Reply by matiasdaloia** [reply]
Thanks! Solved the problem for me :D

**Reply by egbadon-victor** [reply]
It worked!!! Thanks for this

**Reply by msabree** [reply]
Thank you!! What an insane find!

**Reply by DigitalCreationsCo** [reply]
Great job @Pardaramskha!!! Helped me solve.

**Reply by stilllovinnero** [reply]
This helped me! I had similiar issues with Phosphor Icons.

**Reply by alshakero** [reply]
I spent the whole day banging my head into the wall. Thank you so much!

**Comment by OmarFaruk-0x01** [comment]
I am currently using React Tabler Icons (`@tabler/icons-react`) and I have encountered an issue. Initially, everything was working fine, but suddenly our Next.js server was taking too much time to process requests and respond. Can someone please guide me on how to solve this issue?
